### PR TITLE
Prevent generation on non-AST union types

### PR DESCRIPTION
**Pull request by msujew**
Closes https://github.com/langium/langium/issues/941
Closes https://github.com/langium/langium/issues/957
Closes https://github.com/langium/langium/issues/958

### CODE REVIEW

**Review by sailingKieler — CHANGES_REQUESTED** [review]
Great, thanks @msujew for the prompt reaction.

However, 2 wishes:
 * add a bit of prose to the definition of `AstTypeList` & `ValiditionChecks` on what it is supposed to describe/denote (daniel and me tried to guess it yesterday)
 * create at least a related unit test

**Comment by msujew** [comment]
We don't really have a testing infrastructure in place for type testing. Since this is a type change (it doesn't actually change any runtime code), there's little need for runtime tests.

File: packages/langium/src/grammar/type-system/types-util.ts [hunk 1]
```diff
@@ -194,3 +194,18 @@ export function findReferenceTypes(type: PropertyType): string[] {
     }
     return [];
 }
+
+export function hasAstType(type: PropertyType): boolean {
+    if (isPropertyUnion(type)) {
+        return type.types.some(hasAstType);
```
Comment (spoenemann) [inline packages/langium/src/grammar/type-system/types-util.ts:200]: Should we better use `every`? A union that combines AST types with primitives is not usable as AST type.

**Review by spoenemann — COMMENTED** [review]
I don't see #958 fixed, the invalid switch case in `computeIsSubtype` is still generated.

**Review by sailingKieler — APPROVED** [review]
I basically agree, although I would love to see some comments here and there simplifying to understand what's going on and what aim of the particular filter criteria is (e.g. `hasAstType(e.type)`)

Provided a few suggestions for improving the documentation

**Comment by spoenemann** [comment]
AFAICT https://github.com/langium/langium/pull/942#pullrequestreview-1315362762 is still valid.

**Comment by msujew** [comment]
@spoenemann Indeed, I pushed a fix for that. Thanks 👍

**Review by spoenemann — APPROVED** [review]
:+1:

**Pull request merged**